### PR TITLE
Run dokken-validate only for changes to prod code

### PR DESCRIPTION
**Pull request by awslitvit**
### Description of changes
* [Run dokken-validate and system tests only for changes to prod code](https://github.com/aws/aws-parallelcluster-cookbook/pull/1817/commits/a5962c83ebfe2c2e9e8095cb75c5f719d7584590)
* [First stab at ChefSpec tests for aws-parallelcluster-config](https://github.com/aws/aws-parallelcluster-cookbook/pull/1817/commits/45b14cd24181086d57e505c3b9a2673e1f4bb665)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/dokken-validate.yml [hunk 1]
```diff
@@ -23,9 +23,19 @@ jobs:
...
+      - name: Get changed files
+        id: changed-files-excluding-tests
+        uses: tj-actions/changed-files@v35
+        with:
+          files_ignore: |
+            !.github
+            !**/aws-parallelcluster-*/spec
+            !**/aws-parallelcluster-*/test
       - name: Install Chef
+        if: steps.changed-files-excluding-tests.outputs.any_changed == 'true'
```
Comment (psacc) [inline .github/workflows/dokken-validate.yml:35]: redundant, will hit cache, won't it?

File: .github/workflows/dokken-validate.yml [hunk 2]
```diff
@@ -23,9 +23,19 @@ jobs:
     steps:
       - name: Check out code
         uses: actions/checkout@main
+      - name: Get changed files
```
Comment (eantonin) [inline .github/workflows/dokken-validate.yml:26]: nitpick, is there a way ti aviod the duplication of this action in the workflow?

**Review by eantonin — DISMISSED** [review]
LGTM

File: .github/workflows/system-tests-centos7.yml [hunk 3]
```diff
@@ -21,12 +21,22 @@ jobs:
           - 5000:5000
     steps:
       - uses: actions/checkout@main
+      - name: Get changed files
+        id: changed-files-excluding-tests
+        uses: tj-actions/changed-files@v35
+        with:
+          files_ignore: |
```
Comment (psacc) [inline .github/workflows/system-tests-centos7.yml:28]: ignored files are repeated, could you use a variabile to avoid it?

File: .github/workflows/system-tests-centos7.yml [hunk 4]
```diff
@@ -21,12 +21,22 @@ jobs:
...
+            !**/aws-parallelcluster-*/spec
+            !**/aws-parallelcluster-*/test
       - name: Set up Docker Buildx
+        if: steps.changed-files-excluding-tests.outputs.any_changed == 'true'
         id: buildx
         uses: docker/setup-buildx-action@v1
         with:
           driver-opts: network=host
       - name: Install
+        if: steps.changed-files-excluding-tests.outputs.any_changed == 'true'
```
Comment (psacc) [inline .github/workflows/system-tests-centos7.yml:39]: redundant, will hit cache, won't it?

File: .github/workflows/system-tests-centos7.yml [hunk 5]
```diff
@@ -21,12 +21,22 @@ jobs:
...
+      - name: Get changed files
+        id: changed-files-excluding-tests
+        uses: tj-actions/changed-files@v35
+        with:
+          files_ignore: |
+            !.github
+            !**/aws-parallelcluster-*/spec
+            !**/aws-parallelcluster-*/test
       - name: Set up Docker Buildx
+        if: steps.changed-files-excluding-tests.outputs.any_changed == 'true'
```
Comment (psacc) [inline .github/workflows/system-tests-centos7.yml:33]: redundant, will hit cache, won't it?

File: .github/workflows/system-tests-ubuntu.yml [hunk 6]
```diff
@@ -21,12 +21,22 @@ jobs:
...
+      - name: Get changed files
+        id: changed-files-excluding-tests
+        uses: tj-actions/changed-files@v35
+        with:
+          files_ignore: |
+            !.github
+            !**/aws-parallelcluster-*/spec
+            !**/aws-parallelcluster-*/test
       - name: Set up Docker Buildx
+        if: steps.changed-files-excluding-tests.outputs.any_changed == 'true'
```
Comment (psacc) [inline .github/workflows/system-tests-ubuntu.yml:33]: redundant, will hit cache, won't it?

File: .github/workflows/system-tests-ubuntu.yml [hunk 7]
```diff
@@ -21,12 +21,22 @@ jobs:
...
         id: buildx
         uses: docker/setup-buildx-action@v1
         with:
           driver-opts: network=host
       - name: Install
+        if: steps.changed-files-excluding-tests.outputs.any_changed == 'true'
```
Comment (psacc) [inline .github/workflows/system-tests-ubuntu.yml:39]: redundant, will hit cache, won't it?

File: .github/workflows/system-tests-ubuntu.yml [hunk 8]
```diff
@@ -21,12 +21,22 @@ jobs:
           - 5000:5000
     steps:
       - uses: actions/checkout@main
+      - name: Get changed files
+        id: changed-files-excluding-tests
+        uses: tj-actions/changed-files@v35
+        with:
+          files_ignore: |
```
Comment (psacc) [inline .github/workflows/system-tests-ubuntu.yml:28]: ignored files are repeated, could you use a variabile to avoid it?

**Review by psacc — DISMISSED** [review]
I'm ok with the improvement, added just some DRY comments and cache population suggestions

File: .github/workflows/dokken-validate.yml (hunk 2 above)
Comment (awslitvit) [reply .github/workflows/dokken-validate.yml:26]: I couldn't find a solution quickly, we can improve in the future.

File: .github/workflows/system-tests-centos7.yml (hunk 3 above)
Comment (awslitvit) [reply .github/workflows/system-tests-centos7.yml:28]: As far as I know, there is no way to define a variable in the code and share it across workflows.
However, this is just the first stab, we can improve later.

File: .github/workflows/dokken-validate.yml (hunk 1 above)
Comment (awslitvit) [reply .github/workflows/dokken-validate.yml:35]: As agreed offline, we'll monitor workflow duration and remove this `if` in case of negative impact.

File: .github/workflows/system-tests-centos7.yml (hunk 5 above)
Comment (awslitvit) [reply .github/workflows/system-tests-centos7.yml:33]: Same as above.

File: .github/workflows/system-tests-centos7.yml (hunk 4 above)
Comment (awslitvit) [reply .github/workflows/system-tests-centos7.yml:39]: Same as above.

File: .github/workflows/system-tests-ubuntu.yml (hunk 8 above)
Comment (awslitvit) [reply .github/workflows/system-tests-ubuntu.yml:28]: Same as above.

File: .github/workflows/system-tests-ubuntu.yml (hunk 6 above)
Comment (awslitvit) [reply .github/workflows/system-tests-ubuntu.yml:33]: Same as above.

File: .github/workflows/system-tests-ubuntu.yml (hunk 7 above)
Comment (awslitvit) [reply .github/workflows/system-tests-ubuntu.yml:39]: Same as above.

File: .github/workflows/dokken-validate.yml [hunk 9]
```diff
@@ -23,9 +23,19 @@ jobs:
...
+        id: changed-files-excluding-tests
+        uses: tj-actions/changed-files@v35
+        with:
+          files_ignore: |
+            !.github
+            !**/aws-parallelcluster-*/spec
```
Comment (enrico-usai) [inline .github/workflows/dokken-validate.yml:32]: If I understood correctly we're skipping to execute the validate if there are changes in the test code.

This means that if the tests are broken and we're adding a commit to fix them they won't be executed, so won't have confirmation about the fix. Am I correct?

Comment (awslitvit) [reply .github/workflows/dokken-validate.yml:32]: Correct, if the fix contains only the test code. GH Action is required to certify changes to production code: you can always test changes to your test locally.
Furthermore, we can change these conditions in the future if we find out they don't cover correctly some corner cases.

**Review by enrico-usai — APPROVED** [review]
